### ISSUE TITLE
Backup EntryProcessor returns null from entry.getValue() after data rebalancing

### ISSUE THREAD

**Issue by fschuh** [hazelcast/hazelcast#6162]
I found that backup EntryProcessors will return `null` from entry.getValue() for a while after data rebalancing causes the target key to move to a new node.
After several seconds the backup EntryProcessor eventually starts returning the right value from entry.getValue() again, but it took up to 30 seconds in my tests for it to stabilize.

The sample code below calls the same EntryProcessor multiple times on the same key.
If you start multiple nodes and kill one of them off to trigger a rebalance, the issue will manifest itself.
The EntryProcessor keeps running in a loop, and you will see NullPointerExceptions happening on the backup EntryProcessor until eventually entry.getValue() starts returning the correct value again, and the NullPointerExceptions will finally stop.
I found that 3 nodes is a very good number to reproduce this issue, with a reproduction rate of 100% if the killed node is either the primary or backup owner of the key.

This issue happens on Hazelcast 3.5.2 and is very easy to reproduce with the code below:

	public class TestEntryProcessorReliabilityHazelcastRaw {
		static int NUM_MEMBERS = 2;
		static HazelcastInstance hazelcastInstance;
		static Member member;

		static class TestPojo implements Serializable {
			public String name;
			public int value;
		}

		public interface EntryProcessorFunction<K, V> extends Serializable {
			Object process(Map.Entry<K, V> entry);
		}

		static class TestEntryProcessor<K, V> implements com.hazelcast.map.EntryProcessor<K, V>, EntryBackupProcessor<K, V>, Serializable {
			private final EntryProcessorFunction<K, V> function;

			public TestEntryProcessor(EntryProcessorFunction<K, V> function) {
				this.function = function;
			}

			@Override
			public Object process(Map.Entry<K, V> entry) {
				System.out.println("Processing Primary");
				return function.process(entry);
			}

			@Override
			public void processBackup(Map.Entry<K, V> entry) {
				System.out.println("Processing Backup");
				function.process(entry);
			}

			@Override
			public EntryBackupProcessor<K, V> getBackupProcessor() {
				return this;
			}
		}

		public static void main(String[] s) throws Exception {
			setUp();

			// only the master should run test()
			if(member.getUuid().equals(hazelcastInstance.getMap("master").get("master"))) {
				System.out.println("I'm the master, I will do the work");
				try {
					// wait a bit before starting
					Thread.sleep(2500);
				}
				catch(Exception e) {
					e.printStackTrace();
				}
				test();
			}
			else {
				// not the master... do nothing
				System.out.println("I'm not the master, I must stay idle");
			}

			// wait forever
			try {
				Object monitor = new Object();
				synchronized(monitor) {
					monitor.wait();
				}
			}
			catch(Exception e) {
				e.printStackTrace();
			}
		}

		public static void setUp() throws Exception {
			System.out.println("Node starting...");

			Config config = new ClasspathXmlConfig("hazelcast_test.xml");
			hazelcastInstance = Hazelcast.newHazelcastInstance(config);

			member = hazelcastInstance.getCluster().getLocalMember();
			// try to make myself the master
			hazelcastInstance.getMap("master").putIfAbsent("master", member.getUuid());

			CountDownLatch latch = new CountDownLatch(1);

			System.out.println("Node " + member.getUuid() + " waiting for others to join... total members: " + hazelcastInstance.getCluster().getMembers().size());

			hazelcastInstance.getCluster().addMembershipListener(new MembershipListener() {
				@Override
				public void memberAdded(MembershipEvent membershipEvent) {
					// wait for other nodes to join, then release the latch
					if (membershipEvent.getMembers().size() == NUM_MEMBERS) {
						latch.countDown();
					}
				}

				@Override
				public void memberRemoved(MembershipEvent membershipEvent) {}

				@Override
				public void memberAttributeChanged(MemberAttributeEvent memberAttributeEvent) {}
			});

			if(hazelcastInstance.getCluster().getMembers().size() >= NUM_MEMBERS)
				latch.countDown();

			// wait for other nodes to join
			try {
				latch.await();
			}
			catch(Exception e) {}

			System.out.println("Node " + hazelcastInstance.getName() + " started.");
		}

		public static void test() {
			TestPojo pojo = new TestPojo();
			pojo.name = "name";
			pojo.value = 1;

			IMap<String, TestPojo> map = hazelcastInstance.getMap("test_one_backup");

			// put
			map.put("key", pojo);

			//ICompletableFuture<?> future = (ICompletableFuture<?>)map.submitToKey("key", new TestEntryProcessor<String, TestPojo>(entry -> {

			TestEntryProcessor entryProcessor = new TestEntryProcessor<String, TestPojo>(entry -> {
				// update entry
				TestPojo v = entry.getValue();
				int oldValue = v.value;
				v.value++;

				System.out.println("Updating value: " + oldValue + " to " + v.value);

				System.out.println("invoke waiting...");
				try {
					// wait a bit before moving on
					Thread.sleep(5000);
				}
				catch(Exception e) {
					e.printStackTrace();
				}

				System.out.println("finished.");

				// call setValue()
				entry.setValue(v);

				return entry.getValue().value;
			});

			// run once
			Object result = map.executeOnKey("key", entryProcessor);

			try {
				// wait a bit before moving on
				Thread.sleep(5000);
			}
			catch(Exception e) {
				e.printStackTrace();
			}

			// run in a loop
			for(;;) {
				result = map.executeOnKey("key", entryProcessor);
			}

		}

	}

The test map has a synchronous backup count of 1, here's the config.

		<map name="test">
			
			<in-memory-format>BINARY</in-memory-format>

			
			<backup-count>1</backup-count>

		</map>

Note: when testing the code, do not kill the first node that's started, as that's the only node that runs the EntryProcessors.

**Comment by ahmetmircik** [comment]
Hi @fschuh, seems it is an already known issue, did you see this [javaDoc] (https://github.com/hazelcast/hazelcast/blob/v3.5.2/hazelcast/src/main/java/com/hazelcast/map/EntryBackupProcessor.java#L25)?

**Comment by fschuh** [comment]
Hi, thanks for the javadoc link. 
I see, so we have to explicitly deal with this issue then.
I find this misleading because the docs encourage us to use AbstractEntryProcessor, which just runs the same logic in both the primary and in the backup.
If this is a know issue, we can definitely NOT use AbstractEntryProcessor safely.

We can't just deal with a null value the same way in both the primary and in the backup logic, because a null value might mean the data is not yet loaded from the database, for instance.
In this example, this would cause the backup to initialize a fresh new object, while the primary would just update the object that was already in memory.

This is also misleading because EntryProcessors can be run on null keys (key does not yet exist on the map), so you can never be sure what null means.

So the suggestion is to have a slightly different logic between the backup and the primary EntryProcessor when it comes to dealing with nulls?

**Comment by ahmetmircik** [comment]
If there is no data-loss, it can be seen safe to use an `AbstractEntryProcessor`. For now, using slightly different logic when coding backup-processors makes sense to me.

**Comment by fschuh** [comment]
Could this be mentioned in the docs then? I see it as a very dangerous pitfall for anyone using EntryProcessors, especially if they take the easy route and decide to use AbstractEntryProcessor for automatically running the same logic on the backup.

If you say it's safe as long as there's no data loss, this is very important information and should be clearly stated and easy to find in the documentation so there are no surprises.

**Comment by ahmetmircik** [comment]
Sure, want to send a PR? Otherwise i can handle it.

**Comment by fschuh** [comment]
I could do a PR, but I think I can only change the Javadocs, correct?
This information also belongs on the main Hazelcast doc section on EntryProcessors, can that be changed through a PR?

**Comment by ahmetmircik** [comment]
No, you can also send a doc PR. It is in another repo here: https://github.com/hazelcast/hazelcast-reference-manual/blob/master/src/EntryProcessor.md

**Comment by fschuh** [comment]
Great! I'll make the change and send a PR soon.

**Comment by ahmetmircik** [comment]
Closing this issue, PR was already merged. You can find it here https://github.com/hazelcast/hazelcast-reference-manual/pull/41

**Comment by wimdeblauwe** [comment]
I find this issue very interesting. I have overlooked this in the docs. Is there an example somewhere on how to properly code a backup entry processor to avoid this issue? What should the code do when entry.getValue() returns null? Does it just need to do nothing and wait until it gets called again until entry.getValue() is no longer null?

**Comment by fschuh** [comment]
@wimdeblauwe I haven't found any examples of backup processors that deal with this issue, but what I'm doing right now is this:
- Run primary EntryProcessor, and set a flag if entry.getValue() is not null
- On processBackup(), first check whether entry.getValue() returns null
  - If it does return null, and the primary EntryProcessor flag was previously set to true, then skip the backup processor altogether

(*note that this solution requires the flag to be serialized along with the backup processor, or it won't be visible on the backup node. I just make the primary and backup processors the same object, so everything gets serialized together*)
The point here is to skip the backup processor **only** if it receives null when the primary processor doesn't. It will eventually be called again after the data finishes rebalancing properly.

In my humble opinion though, Hazelcast should provide assistance to avoid having us code this behavior in every EntryProcessor we write. Hopefully this will change in the future.

**Comment by wimdeblauwe** [comment]
"In my humble opinion though, Hazelcast should provide assistance to avoid having us code this behavior in every EntryProcessor we write. Hopefully this will change in the future."

I totally agree. Should we open an issue for this?

**Comment by fschuh** [comment]
That was the point of this issue, though it just resulted in a small amendment to the docs.
I think you're right, we should open a new issue targeting the extra amount of work this should be causing to any serious use of EntryProcessors.
It should help make writing them much less cumbersome.

**Comment by wimdeblauwe** [comment]
Ok, can you do it?

**Comment by wimdeblauwe** [comment]
Never mind. I just did it. See #6578

**Comment by fschuh** [comment]
Sorry was a bit busy, you beat me on creating the issue.
Thanks for that, let's hope it gets some follow up.

**Comment by ChristerF** [comment]
This Hazelcast behavior lends me to believe that the only safe contents of a BackupEntryProcessor is to transport the resulting value of the primary EntryProcessor - which kind of makes the concept of a BackupEntryProcessor useless.

**Comment by ChristerF** [comment]
@wimdeblauwe if you skip executing the backup processor, the value between primary and backup will diverge, right? That seems terribly undesirable IMHO.

**Comment by wimdeblauwe** [comment]
Yes, if you change something to the entry in the entry processor, you also need to change the same thing in the backup processor. 

You can read some more info on 2 blog entries I wrote (Note: I am not associated with Hazelcast, so the blog entries just show what I have found out from testing with entry processors):
* https://wimdeblauwe.wordpress.com/2015/09/29/entryprocessors-and-entrybackupprocessors-with-hazelcast/
* https://wimdeblauwe.wordpress.com/2015/09/30/read-only-entryprocessors-with-hazelcast/
